### DISCUSSION TITLE
Getting dependencies from loaded scenes in unity

### DISCUSSION THREAD

**Discussion by Antoshidza** [gustavopsantos/Reflex#47] [Q&A] [answered]
I'm designing a simple mobile unity app. I want to have start scene `AppStartScene` and `AppController` with general for app logic (guess most devs see that solution as a common). Then, for example, after `AppStartScene` and some loading scene, which contains no dependencies, I want to load `MainMenuScene` where I have `MainMenuController` and `MainMenuView`. I want my `AppController` to get dependency of `MainMenuController` after scene is loaded.

So basically my question is: how can I obtain lately registered dependencies for once resolved instance?

**Comment by gustavopsantos** [comment]
At `AppStartScene` create all the dependencies you want for your `MainMenuScene` scene before even loading `MainMenuScene`, and then, use the following instructions to add the dependencies you just instantiated to the container.

```cs
var scene = SceneManager.LoadScene("MainMenuScene", new LoadSceneParameters(LoadSceneMode.Single));
ReflexSceneManager.PreInstallScene(scene, builder => builder.AddSingleton(MainMenuControllerInstance and MainMenuViewInstance));
```

**Reply by Antoshidza** [reply]
As I understand in this example we gonna inject dependencies into `MainMenuScene` scene after load it. But what If I have some complex hierarchy which I don't want to have as a prefabs but do want to inject back to `AppController`?

Asking because `MainMenuView` instance is inside `MainMenuScene` scene, otherwise there is no need in scene if I would use prefabs to show main menu.

**Reply by gustavopsantos** [reply, marked as answer]
So it seems in your case it depends on your implementation. Basically the container heirarchy is almost fully flat, except that all scene container inherits from the project container. Knowing that, "dependencies" on scene containers can depend on "dependencies" from project container, but the opposite is not true.

Container Hierarchy Example:
```mermaid
  graph TD;
      ProjectContainer-->SceneOneContainer;
      ProjectContainer-->SceneTwoContainer;
      ProjectContainer-->SceneThreeContainer;
```

**Reply by Antoshidza** [reply]
Ok, I see, so if I want my `AppController` react to something, for example pressing Exit button, I rather need to implement some `IExitHandler` on `AppController` and just broadcast signal from exit point.

**Reply by Antoshidza** [reply]
What you've used to attach interactive diagram?

**Reply by gustavopsantos** [reply]
@Antoshidza 
I used mermaid, heres the docs https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-diagrams
